### DISCUSSION TITLE
How to write/register truly native components?

### DISCUSSION THREAD

**Discussion by coderedart** [slint-ui/slint#5315] [Q&A] [answered]
At the moment, slint components are just slint files. It can only hold slint values like color/brush/modelRc/string etc.. And the only way components can render themselves is by using Text or Rectangle or Path components.

But how would I write a widget like video player? To use qml as a reference, the VideoOutput component contains a [video sink](https://doc.qt.io/qt-6/qml-qtmultimedia-videooutput.html#videoSink-prop) property which holds a native c++ [QVideoSink](https://doc.qt.io/qt-6/qvideosink.html) object. 

I think qt can register QObjects as QML objects, and they are exposed to qml as any other Component. And when the object is destroyed in qml, it simply calls the native object's destructor too. I want to do something similar for slint components.

In particular, I want to use skia and expose more drawing primitives (eg: skia's backdrop blur shaders or paint object). I can *probably* just set some native callbacks in a global for common drawing routines, But, I cannot figure out how to make a component own some native object (like skia's paint) and pass it around as parameter to these [native callbacks](https://releases.slint.dev/1.5.1/docs/rust/slint_interpreter/struct.componentinstance#method.set_global_callback)

**Comment by ghost** [comment]
Hi @coderedart, 
thank you for your feedback. We have discussed about how to make it possible to also write widget/component code with native code like rust. @tronical had worked on a prototype. If I remember right he want to continue to work on this topic soon. 

I meantime to help you with your video player. There is an example to display a video in Slint https://github.com/slint-ui/slint/tree/master/examples/ffmpeg. This way it is difficult to create a truly shareable video widget, but it could be a start.

And for the custom drawing with skia part. What works always is to draw something, create an `Image` object from native and set it as `source` of a Slint `Image` component. I have experimented with tiny-skia this way and it worked well for me.

**Reply by coderedart** [reply]
Thanks for the quick reply. the ffmpeg example is a great reference. With some hacks, I might be able to get multiple video playback working. 

I do have a followup question. I am dealing with a plugin system, and am thinking of dynamically loading plugin's slint files at runtime with interpretor and placing them in my host application's window. Is that possible at the moment? 

If those plugin's slint files, wanted to use my custom VideoPlayer component, I would need to give them some "global" callbacks too, which they can use to toggle video playback inside their components. 

When the plugins are unloaded, will the global callbacks of their components be dropped? As I would need to clean up the player instances when the plugin's UI is closed.

**Reply by ghost** [reply, marked as answer]
Currently to provide pre-compiled Slint components e.g. in a separate crate and use them in you application project does not work or am I wrong @tronical? What you can do is to ship slint files with your plugin you can use my custom widget crate as example

* https://flovansl.codeberg.page/coop_sl/snapshots/main/docs/rust/coop/
* https://codeberg.org/flovansl/co_sl/src/branch/main/widgets/coop/src/lib.rs

What you also can do is to ship generic functionality that is based on types from the slint-ui crate. For example I have done an generic keyboard selection behavior for models in Slint https://codeberg.org/flovansl/co_sl/src/branch/main/apps/coop_local/rust/src/item_selector.rs

**Reply by coderedart** [reply]
Thank you for the answers. It seems too much effort to wrangle all of this for my use case. I will just wait for the native widget work by @tronical .

**Comment by renlite** [comment]
"@tronical had worked on a prototype. If I remember right he want to continue to work on this topic soon."
Where could I get more information about writing widgets/components in rust.

As far as I know Slint compiler creates Rust code that is compiled to native. Is this correct? In opposite the Python-API can work with interpreted Slint-Code only.

**Reply by ghost** [reply]
Right now Slint code can compiles to Rust and C++. The Phyton and JS/TS version are using the  interpreter.  A Python compiler is something that is possible but not planned for the near future.
